Prioritize reentrant tasks and execute them recursively until close to limit

malvika2147 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22397 Prioritize reentrant tasks and execute them recursively until close to limit**

Summary: The ready queue prioritizes the most nested reentrant tasks. Run the reentrant tasks recursively until the recursion depth is close to max_recursion_depth, which depends on the python recursion limit. Once the limit is reached, further reentrant backwards tasks will be run in a different thread. 

Test Plan: Added test for reentrant backwards with checkpoint and a test for a recursive backwards function (which should fail if we run all the reentrant tasks recursively in the same thread) and for testing priority of reentrant tasks.
~~Will add a test for priority of reentrant tasks in future pr.~~

Differential Revision: [D16131955](https://our.internmc.facebook.com/intern/diff/D16131955)

Commit:
Prioritize reentrant tasks and execute them recursively until close to limit

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

ezyang commented on torch/csrc/autograd/python_engine.cpp:
```diff
@@ -31,6 +32,10 @@ static Engine& get_python_engine() {
 namespace torch { namespace autograd { namespace python {
 
+PythonEngine::PythonEngine () {
+  max_recursion_depth_ = 0.1*Py_GetRecursionLimit();
```
A little note explaining how we got the magic number 0.1 would be appreciated here :)

ezyang replied:
So, I guess, now the constructor for PythonEngine must be GIL protected. Have we checked this invariant is respected in all the use sites?

ezyang commented:
> Test Plan

A formally written out test plan would be very useful here, since there are no test suite changes (and indeed, it's a bit hard to think of how you would actually go about writing unit tests; probably only manual tests are possible here.)

ezyang commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -79,10 +84,12 @@ struct CompareFunctionTaskTime {
+    } else if (t1.base_== t2.base_) {
       return t1.fn_->sequence_nr() < t2.fn_->sequence_nr();
+    } else {
+      return t1.getReentrantDepth() <= t2.getReentrantDepth();
```
Hmm. But if the reentrant depths match, don't I still want to test on sequence numbers? So it seems like if there is a difference in reentrant depth, you should use that, otherwise you should defer to sequence number comparison here.

ezyang commented on torch/csrc/autograd/engine.h:
```diff
@@ -69,9 +70,13 @@ struct TORCH_API Engine {
   // Safe to read ready_queues_ without synchronization after intialization
   std::vector<std::shared_ptr<ReadyQueue>> ready_queues_;
+  // Number of reentrant threads currently accessing ready queue
+  std::vector<int> num_reentrant_threads_;
```
Is this protected by a lock?

ezyang replied:
The comment here would be clearer if it said something like, "For each ready queue, how many reentrant threads are currently accessing it."

malvika2147 replied:
We only expect one thread to be accessing a particular ready at a time, so we don't need a lock.

ezyang replied:
:+1: Worth a comment. Fun fact: accessing disjoint indices from different threads is thread safe... unless you are using`std::vector<bool>` (in which case it's not)

ezyang commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -651,15 +665,25 @@ auto Engine::execute(const edge_list& roots,
-    graph_task.not_done_.wait(lock, [&graph_task]{
-      return graph_task.outstanding_tasks_.load() == 0;
-    });
+    graph_task.reentrant_depth_ = num_reentrant_threads_[worker_device+1]*max_recursion_depth_+current_depth;
```
Oh interesting. I guess the reason you have written the code this way, is that when you make a new worker thread, you "lose" the setting of `current_depth_`. So to get around this problem, you maintain `num_reentrant_threads_` to remember the value for the engine.

I think there is a simpler way to solve this problem though! Since every graph task has a reentrant depth, the computation here should just be `graph_task.reentrant_depth_ = reentrant_depth_ + 1`. Am I mistaken about this? In particular, it doesn't feel like the tracking in `num_reentrant_threads_` should be necessary. I haven't thought about this in as much detail as you, so please let me know if there is a good reason here.

malvika2147 replied:
We are not keeping track of the graph task while executing the FunctionTask function. So I don't think it'll be available in Engine::execute.

Commit:
Update on "Prioritize reentrant tasks and execute them recursively until close to limit"

Prioritize reentrant tasks and execute them recursively until close to limit

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

gh-metadata: pytorch pytorch 22397 gh/mal2147/10/head

Commit:
Update on "Prioritize reentrant tasks and execute them recursively until close to limit"

Prioritize reentrant tasks and execute them recursively until close to limit

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

gh-metadata: pytorch pytorch 22397 gh/mal2147/10/head

ezyang commented:
Don't forget to rerequest review when you need it!

ezyang commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -342,7 +358,7 @@ auto Engine::thread_main(GraphTask *graph_task) -> void {
   }
 
-  if (graph_task) {
+  if (graph_task && current_depth == 0) {
```
This line made me have to pause and think for a moment. It would have been a bit clearer if it said something in a comment like, "If `current_depth == 0`, then this worker thread is done working, and we need to suspend it and transfer control back to the original thread."

This also makes me wonder about an edge case, where you can trigger pathological behavior if you repeatedly make a reentrant call and then exit *right* on one of these "stack chunk" boundaries. Nothing to fix in this PR, but might be worth noting down somewhere in code.

Commit:
Update on "Prioritize reentrant tasks and execute them recursively until close to limit"

Prioritize reentrant tasks and execute them recursively until close to limit

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

gh-metadata: pytorch pytorch 22397 gh/mal2147/10/head

Commit:
Update on "Prioritize reentrant tasks and execute them recursively until close to limit"

Prioritize reentrant tasks and execute them recursively until close to limit

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

gh-metadata: pytorch pytorch 22397 gh/mal2147/10/head

Commit:
Update on "Prioritize reentrant tasks and execute them recursively until close to limit"

Prioritize reentrant tasks and execute them recursively until close to limit

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

gh-metadata: pytorch pytorch 22397 gh/mal2147/10/head

Commit:
Update on "Prioritize reentrant tasks and execute them recursively until close to limit"

Prioritize reentrant tasks and execute them recursively until close to limit

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

gh-metadata: pytorch pytorch 22397 gh/mal2147/10/head

VitalyFedyunin commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -651,15 +674,25 @@ auto Engine::execute(const edge_list& roots,
+    if(current_depth >= max_recursion_depth_){
+      // See Note [Reentrant backwards]
+      // If reached the max depth, switch to a different thread
+      add_thread_pool_task(&graph_task);
```
What will happen if we exceed thread_pool_shared_->graphtasks_queue_.size() here?

malvika2147 replied:
std::queue should resize dynamically, or am I misunderstanding your question?

VitalyFedyunin commented:
I suggest to have clearly defined constant how deep we can go with recursion.

malvika2147 commented:
> I suggest to have clearly defined constant how deep we can go with recursion.

But in that case, we will always create new threads even if there is enough stack space to run on a single thread.

Commit:
Update on "Prioritize reentrant tasks and execute them recursively until close to limit"

Prioritize reentrant tasks and execute them recursively until close to limit

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

gh-metadata: pytorch pytorch 22397 gh/mal2147/10/head

ezyang commented on test/test_autograd.py:
```diff
@@ -3219,6 +3219,50 @@ def backward(ctx, x):
+                with torch.enable_grad():
+                    ctx.x = Variable(x.data, requires_grad=True)
+                    ctx.x = ctx.x - 1
+                return ctx.x.detach()
```
This is a pretty funky forward function. (It's funky because we don't normally do autograd operations inside of a forward function.) It seems like it's doing two things: you want to `return x` (forward is just identity), but you also want to create a leaf variable on context with some non-trivial autograd history. Is there a reason `x` has to be used in both cases? I'll keep reading and see if I can figure out why you create leaf variables in forward ;)

ezyang replied:
Oh I see, you're also using `ctx.x` to keep track about how many times you recurse.

ezyang commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -186,19 +186,19 @@ struct GraphTask {
   }
 
-  GraphTask(bool keep_graph, bool grad_mode)
+  GraphTask(bool keep_graph, bool grad_mode, int reentrant_depth = 0)
```
Are you sure you want this to be an optional argument? It might hide some bugs where you need to pass in reentrant depth but you forget because it defaults to 0.

Commit:
Update on "Prioritize reentrant tasks and execute them recursively until close to limit"

Prioritize reentrant tasks and execute them recursively until close to limit

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

gh-metadata: pytorch pytorch 22397 gh/mal2147/10/head

mrshenli commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -79,10 +86,12 @@ struct CompareFunctionTaskTime {
+      return t1.getReentrantDepth() < t2.getReentrantDepth();
     }
-  }
+   }
```
indent?

mrshenli commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -216,7 +232,8 @@ auto ReadyQueue::pop() -> FunctionTask {
 }
 
-Engine::Engine() = default;
+// This limit is based on the default python recursion limit which is 1000
```
comments do not match the code 1000 vs 100

malvika2147 replied:
We need to set it lower than the actual python limit to take into account the function calls within python

Commit:
Update on "Prioritize reentrant tasks and execute them recursively until close to limit"

Prioritize reentrant tasks and execute them recursively until close to limit

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

gh-metadata: pytorch pytorch 22397 gh/mal2147/10/head
